mlir/lib/Dialect/Rock/Transforms/SugarToLoops.cpp

```diff
@@ -508,7 +599,15 @@ struct IndexDiffUpdateRewritePattern
           return b.create<ConstantIndexOp>(loc, diff + mbOriginalConst.value());
         }
       }
-      return b.create<AddIOp>(loc, original, diff);
+      // If the operation is something like Embed{} or Pad{} that can produce
+      // negative results (but thereby cancel validity - that is, if a negative
+      // result means all the further operations are being speculatively
+      // executed), we conservatively don't say anything. However, in other
+      // cases, the original value must be non-negative and the index diff can't
+      // make it go below zero, so we can declare nuw.
+      return b.create<AddIOp>(loc, original, diff,
+                              canBeInvalid ? IntegerOverflowFlags::none
+                                           : (IntegerOverflowFlags::nuw));
```

Comment:
Why is the discrepency here to not add nsw as well as compared affine map expanders ?

Reply:
I think `5 + -1` isn't `nsw`, which can happen in index diffs and be valid but can't happen validly in the affine expansion

Reply:
makes sense.